addons/product/models/product_attribute_value.py

```diff
@@ -98,26 +93,35 @@ def write(self, values):
             self.env.invalidate_all()
         return res
 
+    def check_is_used_on_products(self):
+        for pav in self.filtered('is_used_on_products'):
+            return _(
+                "You cannot delete the value %(value)s because it is used on the following"
+                " products:\n%(products)s\n",
+                value=pav.display_name,
+                products=", ".join(pav.pav_attribute_line_ids.product_tmpl_id.mapped('display_name')),
+            )
+        return False
+
     @api.ondelete(at_uninstall=False)
     def _unlink_except_used_on_product(self):
+        if is_used_on_products := self.check_is_used_on_products():
+            raise UserError(is_used_on_products)
+
+    def unlink(self):
+        archived_pavs = self.env['product.attribute.value']
         for pav in self:
-            if pav.is_used_on_products:
-                raise UserError(_(
-                    "You cannot delete the value %(value)s because it is used on the following "
-                    "products:\n%(products)s\n If the value has been associated to a product in the"
-                    " past, you will not be able to delete it.",
-                    value=pav.display_name,
-                    products=", ".join(pav.pav_attribute_line_ids.product_tmpl_id.mapped('display_name')),
-                ))
             linked_products = pav.env['product.template.attribute.value'].search(
                 [('product_attribute_value_id', '=', pav.id)]
             ).with_context(active_test=False).ptav_product_variant_ids
-            unlinkable_products = linked_products._filter_to_unlink()
-            if linked_products != unlinkable_products:
-                raise UserError(_(
-                    "You cannot delete value %s because it was used in some products.",
-                    pav.display_name
-                ))
+            active_linked_products = linked_products.filtered('active')
+            if not active_linked_products:
+                # If product attribute value found on non-active product variants
+                # archive PAV instead of deleting
+                archived_pavs |= pav
+        if archived_pavs:
+            return archived_pavs.action_archive()
+        return super().unlink()
```

Comment:
There is again problem with multiple returns, so if there are `archived_pavs` it will not propagate the other records of self to the parent function. So I suggest doing smth like that.
```suggestion
        if archived_pavs:
             archived_pavs.action_archive()
        return super(ProductAttributeValue, self - archived_pavs).unlink()
```